src/gwc/src/main/java/org/geoserver/gwc/layer/GeoServerTileLayer.java

```diff
@@ -579,57 +587,303 @@ protected final void sendTileRequestedEvent(ConveyorTile tile) {
         }
     }
 
-    protected ConveyorTile getMetatilingReponse(
-            ConveyorTile tile, final boolean tryCache, final int metaX, final int metaY)
+    protected ConveyorTile getMetatilingResponse(
+            ConveyorTile conveyorTile, final boolean tryCache, final int metaX, final int metaY)
             throws GeoWebCacheException, IOException {
 
-        if (tryCache && tryCacheFetch(tile)) {
-            return finalizeTile(tile);
+        if (tryCache && tryCacheFetch(conveyorTile)) {
+            return finalizeTile(conveyorTile);
         }
 
-        final GeoServerMetaTile metaTile = createMetaTile(tile, metaX, metaY);
-        Lock lock = null;
+        final GeoServerMetaTile metaTile = createMetaTile(conveyorTile, metaX, metaY);
+
+        /* ****************** Acquire lock on metatile ******************* */
+        final Lock metaTileLock =
+                GWC.get().getLockProvider().getLock(buildMetaTileLockKey(conveyorTile, metaTile));
+
         try {
-            /* ****************** Acquire lock ******************* */
-            lock = GWC.get().getLockProvider().getLock(buildLockKey(tile, metaTile));
-            // got the lock on the meta tile, try again
-            if (tryCache && tryCacheFetch(tile)) {
-                LOGGER.finest(
-                        "--> "
-                                + Thread.currentThread().getName()
-                                + " returns cache hit for "
-                                + Arrays.toString(metaTile.getMetaGridPos()));
-            } else {
-                LOGGER.finer(
-                        "--> "
-                                + Thread.currentThread().getName()
-                                + " submitting getMap request for meta grid location "
-                                + Arrays.toString(metaTile.getMetaGridPos())
-                                + " on "
-                                + metaTile);
-                WebMap map;
+            boolean foundInCache = false;
+            if (tryCache) {
+                /* ****************** Acquire lock on individual tile ******************* */
+                // Will block here if there is an async thread currently saving this tile
+                String lockKey = buildTileLockKey(conveyorTile, conveyorTile.getTileIndex());
+                final Lock tileLock = GWC.get().getLockProvider().getLock(lockKey);
```

Comment:
ok so for the on-demand seeding this looks good. At this point we'll hold both the metatile and tile locks.
For the seeding jobds though it looks like we'll be acquiring the tile locks anyway, as called by `SeedTask`: 
```java
boolean tryCache = !reseed
...
tl.seedTile(tile, tryCache);
```
I think we should prevent acquiring the tile lock as in `computeMetatile()` below (`if (Dispatcher.REQUEST.get() == null)` then don't lock the tile)

Reply:
Yep, I've reshuffled code around so that the tile lock is not taken in that situation.